cip/CIP2016-06-14-Redefine-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc

```diff
@@ -0,0 +1,367 @@
+** either `m1.k = m2.k` is `true`,
+** or both `m1.k IS NULL` and `m2.k IS NULL`
+
+This is at odds with the decision to use <<unknown-null,"unknown" `null` values>> in <<comparability-equality>>.
+
+However, this definition is aligned with the most common use case for maps with `null` entries: updating multiple properties through the use of a single `SET` clause, e.g. `SET n += { size: 12, remove_this_key: null }`. In this case, there is no need to differentiate between different `null` values, as `null` merely serves as a marker for keys to be removed (i.e. is a <<missing-null,"missing" `null` value>>). Current equality semantics make it easy to check if two maps would correspond to the same property update in this scenario. We note though that this type of update map comparison is rare and could be emulated using a more complex predicate. The current rules do however break symmetry with how equality handles `null` in all other cases. This becomes more apparent by considering these two examples:
+
+* `expr1 = expr2` evaluates to `null` if `expr1 IS null && expr2 IS NULL`
+* `{a: expr1} = {a: expr2}` evaluates to `true` if `expr1 IS null && expr2 IS NULL`
+
+====== New map equality ======
+
+To rectify this, we propose instead that two maps `m1` and `m2` should be equal if:
+
+* `m1` and `m2` have the same keys,
+** including keys that map to a `null` value (the order of keys as returned by `keys()` does not matter here).
+* Additionally, for each such key `k`,
+** `m1.k = m2.k` is `true`.
+
+As a consequence of these changes, plain <<equality-def>> is not reflexive for all values (consider: `{a: null} = {a: null}`, `[null] = [null]`). However this was already the case (consider: `null = null` \=> `null`).
+
+Note that <<equality-def>> is reflexive for values that do not involve `null` though.
+
+[[orderability-equivalence]]
+=== Orderability and equivalence ===
+
+We propose that <<orderability-def>> and <<equivalence-def>> should be aligned with each other, i.e.
+
+`expr1` is equivalent to `expr2` if and only if they have the same position under orderability (i.e. they would be sorted before (or after resp.) any other non-equivalent value in the same way).
+
+Orderability and equivalence use <<missing-null,"missing" `null` values>>.
+
+[[orderability-def,orderability]]
+==== Orderability ====
+
+[[global-sort-order,global sort order]]
+We propose that orderability be defined between any pair of values, where the result is always `true` or `false`.
+
+To accomplish this, there must be a pre-determined order of types and each value must fall under exactly one disjoint leaf type in this order. We propose to use the following ascending global sort order of disjoint types:
+
+* `MAP` types
+** <<regular-map,Regular map>>
+** `NODE`
+** `RELATIONSHIP`
+* `LIST OF ANY?`
+* `PATH`
+* `STRING`
+* `BOOLEAN`
+* `NUMBER`
+** `NaN` is treated as the largest number in orderability only (i.e. it is put after positive infinity)
```

Comment:
`s/largest/greatest/`, we should (consistently) talk about values being _less than_ or _greater than_. We also shouldn't talk about it being "put after" without specifying whether the order is ascending or descending.